android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/Router.kt

```diff
@@ -0,0 +1,272 @@
+/*
+ * Copyright (C) 2017. Uber Technologies
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.uber.rib.core
+
+import android.os.Looper
+import androidx.annotation.CallSuper
+import androidx.annotation.MainThread
+import androidx.annotation.VisibleForTesting
+import com.uber.rib.core.Rib.Companion.configuration
+import java.util.Locale
+import java.util.concurrent.CopyOnWriteArrayList
+
+/**
+ * Responsible for handling the addition and removal of children routers.
+ *
+ * @param <I> type of interactor this router routes.
+ */
+abstract class Router<I : Any> protected constructor(
```

Comment:
Kotlin enforces the finite bounds restriction which prevents cyclical types. this drops Interactor to the looser non-null `Any`. a future improvement could be to add an interface that only `Interactor` implements and change the type to that. that would be similar to [what iOS does](https://github.com/uber/RIBs/blob/80e6cea7bdb7c06a7b403b7cdfe2dfa7f1138109/ios/RIBs/Classes/Router.swift#L76)

Reply:
Please creating a tracking task for that.

Reply:
#435